index.d.ts

```diff
@@ -1,71 +1,96 @@
+		@default Infinity
+		*/
+		readonly maxAge?: number;
 
-	/**
-	 * Use a different cache storage. Must implement the following methods: `.has(key)`, `.get(key)`, `.set(key, value)`, `.delete(key)`, and optionally `.clear()`. You could for example use a `WeakMap` instead or [`quick-lru`](https://github.com/sindresorhus/quick-lru) for a LRU cache.
-	 *
-	 * @default new Map()
-	 */
-	readonly cache?: CacheStorage<CacheKeyType, ReturnType>;
+		/**
+		Determines the cache key for storing the result based on the function arguments. By default, if there's only one argument and it's a [primitive](https://developer.mozilla.org/en-US/docs/Glossary/Primitive), it's used directly as a key, otherwise it's all the function arguments JSON stringified as an array.
 
-	/**
-	 * Cache rejected promises.
-	 *
-	 * @default false
-	 */
-	readonly cachePromiseRejection?: boolean;
+		You could for example change it to only cache on the first argument `x => JSON.stringify(x)`.
+		*/
+		readonly cacheKey?: (...arguments: ArgumentsType) => CacheKeyType;
+
+		/**
+		Use a different cache storage. You could for example use a `WeakMap` instead or [`quick-lru`](https://github.com/sindresorhus/quick-lru) for a LRU cache.
+
+		@default new Map()
+		*/
+		readonly cache?: CacheStorage<CacheKeyType, ReturnType>;
+
+		/**
+		Cache rejected promises.
+
+		@default false
+		*/
+		readonly cachePromiseRejection?: boolean;
+	}
 }
 
-/**
- * [Memoize](https://en.wikipedia.org/wiki/Memoization) functions - An optimization used to speed up consecutive function calls by caching the result of calls with identical input.
- *
- * @param fn - Function to be memoized.
- */
 declare const mem: {
+	/**
+	[Memoize](https://en.wikipedia.org/wiki/Memoization) functions - An optimization used to speed up consecutive function calls by caching the result of calls with identical input.
+
+	@param fn - Function to be memoized.
+
+	@example
+	```
+	import mem = require('mem');
+
+	let i = 0;
+	const counter = () => ++i;
+	const memoized = mem(counter);
+
+	memoized('foo');
+	//=> 1
+
+	// Cached as it's the same arguments
+	memoized('foo');
+	//=> 1
+
+	// Not cached anymore as the arguments changed
+	memoized('bar');
+	//=> 2
+
+	memoized('bar');
+	//=> 2
+	```
+	*/
 	<
 		ArgumentsType extends unknown[],
 		ReturnType extends unknown,
 		CacheKeyType extends unknown
 	>(
 		fn: (...arguments: ArgumentsType) => ReturnType,
-		options?: Options<ArgumentsType, CacheKeyType, ReturnType>
+		options?: mem.Options<ArgumentsType, CacheKeyType, ReturnType>
 	): (...arguments: ArgumentsType) => ReturnType;
 
 	/**
-	 * Clear all cached data of a memoized function.
-	 *
-	 * @param fn - Memoized function.
-	 */
+	Clear all cached data of a memoized function.
+
+	@param fn - Memoized function.
+	*/
 	clear<ArgumentsType extends unknown[], ReturnType extends unknown>(
 		fn: (...arguments: ArgumentsType) => ReturnType
 	): void;
+
+	// TODO: remove this on next major release
+	default: typeof mem;
```

Comment:
This just happens to work because we have merged the `const mem` declaration with the `namespace mem` declaration. This is more of a bug than a feature. There is a special compiler flag, `esModuleInterop` just to fix this issue. So for users who use this flag, `import mem from 'mem'` won't work without an explicit `default` property on the type of `mem`.

Reply:
I'm not using any flags, I tried in a brand-new project without tsconfig.

Reply:
Why would you disable interop?

Reply:
Sorry, I've just mixed things up.

Following works without `esModuleInterop` because of the merged `const` and `namespace` declarations:
```ts
import * as mem from 'mem';

mem(...);
```

This shouldn't actually work, star imports shouldn't be callable as per ES spec. This is being fixed by the `esModuleInterop` flag:
```
// with esModuleInterop:
import * as mem from 'mem';

mem(...); // Error

// ===================

import mem from 'mem';

mem(...); // Works
```

> This doesn't seem to be required. `import mem from 'mem'` seems to work correctly without it.

This can work only in the following cases:
- you have an explicit `default` property on the thing you export
- you use the `esModuleInterop` which synthesises the `default` export

Reply:
> Why would you disable interop?

I don't think that there is a reason to disable it, it's just that not everyone is aware of this flag (and there are maybe compatibility issues). So dropping the `default` property would be a breaking change.